File: app/src/main/java/com/magdenbt/collectionsbenchmark/ui/viewmodelflow/StatViewModelFactory.java
package com.magdenbt.collectionsbenchmark.ui.viewmodelflow;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.magdenbt.collectionsbenchmark.CollectionsType;
import com.magdenbt.collectionsbenchmark.modelflow.StatRepository;

public class StatViewModelFactory implements ViewModelProvider.Factory {

    private final CollectionsType collectionsType;
    private final StatRepository statRepository;
    private final Application application;


    public StatViewModelFactory(Application application, StatRepository statRepository, CollectionsType collectionsType) {
        this.collectionsType = collectionsType;
        this.statRepository = statRepository;
        this.application = application;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(StatViewModel.class))
            return (T) new StatViewModel(application, statRepository, collectionsType);
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/magdenbt/collectionsbenchmark/ui/viewmodelflow/StatViewModelFactory.kt
package com.magdenbt.collectionsbenchmark.ui.viewmodelflow

import android.content.Context
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class StatViewModelFactory(
    val context: Context,
) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(StatViewModel::class.java)) {
            return StatViewModel(context = context) as T
        }
        Log.e(this.javaClass.canonicalName, "${modelClass.canonicalName} - unknown ViewModel class")
        throw IllegalAccessException("Unknown ViewModel class")
    }
}
